controllers/cassandracluster/storageupsize/action_api.go

```diff
@@ -0,0 +1,95 @@
+package storageupsize
+
+import (
+	"context"
+
+	api "github.com/cscetbon/casskop/api/v2"
+	"github.com/cscetbon/casskop/controllers/cassandracluster/pods"
+	"github.com/cscetbon/casskop/controllers/cassandracluster/storagestateclient"
+	"github.com/cscetbon/casskop/controllers/cassandracluster/sts"
+	"github.com/cscetbon/casskop/controllers/cassandracluster/view"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+)
+
+func ShouldBeStarted(rack view.RackView, requestedCapacity string) bool {
+	requested := silentParseResourceQuantity(requestedCapacity)
+	_, current := findDataCapacity(rack.StoredStatefulSet().Spec.VolumeClaimTemplates)
+	if !requested.Equal(current) {
+		rack.Log().Infof("Storage upsize should be started: ask %v and have %v", requested, current)
+		return true
+	}
+	return false
+}
+
+func Start(rack view.RackView) {
+	startUpsizeAction(rack)
+}
+
+func IsStarted(dcRackStatus *api.CassandraRackStatus) bool {
+	return dcRackStatus.CassandraLastAction.Name == api.ActionStorageUpsize.Name &&
+		dcRackStatus.CassandraLastAction.Status != api.StatusDone
+}
+
+func Reconcile(ctx context.Context, cc *api.CassandraCluster, rack view.RackView,
+	setNewDataCapacity func(statefulSet *appsv1.StatefulSet) error,
+	storageStateClient storagestateclient.StorageStateClient, stsClient sts.StsClient, podsClient pods.PodsClient) error {
+
+	if stepResult := dumpOldStatefulSet(rack); stepResult.HasError() {
+		return stepResult.Error()
+	} else if stepResult.ShouldBreakReconcileLoop() {
+		return nil
+	}
+
+	if stepResult := removeStatefulSetOrphan(ctx, cc, rack, stsClient); stepResult.HasError() {
+		return stepResult.Error()
+	} else if stepResult.ShouldBreakReconcileLoop() {
+		return nil
+	}
+
+	if stepResult := recreateStatefulSetWithNewCapacity(ctx, rack, setNewDataCapacity, stsClient); stepResult.HasError() {
+		return stepResult.Error()
+	} else if stepResult.ShouldBreakReconcileLoop() {
+		return nil
+	}
+
+	dataPVCs, err := getAllDataPvcs(ctx, cc, rack, storageStateClient)
+	if err != nil {
+		return err
+	}
+
+	if stepResult := ensureAllPVCsHaveNewCapacity(ctx, cc, dataPVCs, rack, storageStateClient); stepResult.HasError() {
+		return stepResult.Error()
+	} else if stepResult.ShouldBreakReconcileLoop() {
+		return nil
+	}
+
+	if stepResult := waitTillAllFilesystemsHaveNewCapacity(cc, dataPVCs, rack); stepResult.HasError() {
+		return stepResult.Error()
+	} else if stepResult.ShouldBreakReconcileLoop() {
+		return nil
+	}
+
+	if stepResult := waitTillStatefulSetAndAllPodsAreReady(ctx, cc, rack, podsClient); stepResult.HasError() {
+		return stepResult.Error()
+	}
```

Comment:
Probably makes sense to add `if stepResult.ShouldBreakReconcileLoop()` for consistency or convert it into array of functions called in a loop